### ISSUE TITLE
Granular datum reprocessing

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#3773]
**What is the goal / desired outcome?**

If I've accidentally updated one of my pipelines to contain a bad version of
the code that processes datums and returns a 0 exit code, but doesn't produce
the right result I'd like to be able to reprocess just the datums that were
processed with that version.

**If there is a way to accomplish this today via workaround, what does that require?**

The only ways to solve this are to reprocess everything or to artificial modify
the processed datums so they get rerun.

**(Optional) What is your proposal for a feature to solve this?**

The first thing that has to change is the way we deduplicate datum
processing. Presently we store information about which datums were already
processed in the parent commit AND in object storage using a hash computed
from the pipeline and the hash of the data. I think that we should do away
with this second method and only use the first. This will mean that datums
only get skipped if they were processed in the parent job, which means
there are now a number of ways to reprocess a datum:

- delete the commits that contain datum and recommit them
- create a new commit that deletes the datums you want to reprocess, then
  recommit (probably with copy-file)
- delete the output commit (not yet supported, but could be added as part
  of this, I think it's pretty easy) which should retrigger the job at
  head and reprocess the datums that were processed in the deleted
  commits.
- delete the latest version (or versions) of the pipeline. (also not yet
  supported, but I think very easy to add, it just comes down to deleting
  the spec commit for that version of the pipeline). Then update the
  pipeline to reprocess all the things that were processed by the deleted
  version(s).

Downsides to this approach:

We'll be losing some deduplication functionality by implementing this.
(The above cases would have resulted in a datum being deduplicated.)
However, I think this is desirable, all the cases above are fairly rare,
and I can't really think of a reason to do them other than wanting the
datums to be reprocessed (or not processed at all). So I think for most
users not deduplicating will be less confusing and if you don't want to
pay the costs of reprocessing I think you can just avoid doing these
operations.

Output commit parentage

This also gives a decision to another behavior we've gone back and forth
on, which is when output commits should be parented on each other. The
answer here is that they're parented if the child uses the parent's datums
for dedupe purposes, in fact the parentage is the reason they're deduped.
This means that when you do `update pipeline --reprocess` the parentage
chain should get broken, `update pipeline` without `--reprocess` the
parentage chain should remain in tact.

Migration

Probably the hardest part of this is going to be migration. For one it's
going to involve a change in the format of hashtrees, so they'll probably
have to be recomputed. But, if done in the naive way it's also going to
require reprocessing everything. We'll probably wind up writing some sort
of compatibility hack that uses the partial datum trees one last time to
create the first output commit for each pipeline and merge the hashtree
into the right state.

**Comment by gabrielgrant** [comment]
Am I understanding correctly that, for the specific use-case of wanting to reprocess one specific datum, all those methods will require an extra commit (that deletes the relevant file(s)), and therefore an extra (no-op) job?

**Comment by jdoliner** [comment]
That's correct, which does seem a bit clunky in terms of needing to run the no-op job, but I think the interface is fairly intuitive, in that I've seen people try to do exactly this to reprocess a datum. I'm not sure what a better interface would look like right now.

**Comment by gabrielgrant** [comment]
Yea, no-op job certainly isn't the end of the world, just wanted to make sure i was understanding the implications. Ideally, i think something like an `invalidate-datum` command, that could be issued multiple times within the context of a single transaction, would be the most intuitive. I'd imagine this would do something like incrementing a salt used to compute the datum's hash, or, in the context of the new stuff @brycemcanally is working on, could possibly just directly create a new diff set that gets assigned to a job? I don't understand the details of that stuff well enough to know whether this sort of an implementation makes any sense, so fully I understand such a thing may be challenging to implement.

**Comment by brycemcanally** [comment]
For performance reasons, I agree that skipping should only be based on what we find in the parent commit. I think with diff file sets we will have to re-think what re-processing looks like because we are not computing all of the datum hashes from scratch with that model. A simple solution would be to just not use the diff file set when we need to re-process. As far as re-processing individual datums is concerned, I think that logic should wrap a diff file set rather than be a part of the diff file set.

**Comment by gabrielgrant** [comment]
@brycemcanally sorry, my ignorance of the storage layer is showing through here, but I don't really follow your comment:

- What is the implication of "skipping only being based on what we find in the parent commit"?
- What do you mean by the reprocessing "logic wrapping a diff file set rather than being a part of the diff file set"?

**Comment by brycemcanally** [comment]
> What is the implication of "skipping only being based on what we find in the parent commit"?

That statement is not very precise, but what I mean is that we should skip based on information associated with the parent output commit, not some global notion of datums that have been processed whenever and wherever. The current model has a set of datums associated with an output commit (stored as an object) to indicate the datums processed in the associated job. We use this as a fast path for skipping when we can. In the new model, we will have a file set (in similar form as stats is now #3703) associated with the parent output commit that will be checked to determine which datums to skip. We will first determine the datums we need to check from the input diff file set, then compute their hashes and compare them to what we have stored in the parent output commit's datum file set (refer to #3703).

> What do you mean by the reprocessing "logic wrapping a diff file set rather than being a part of the diff file set"?

I am mainly addressing this comment

> could possibly just directly create a new diff set that gets assigned to a job?

with that statement. I interpreted that as meaning we would have re-processing information (new salt or whatever) stored in the diff file set. The diff file sets, in my opinion, should just have pfs information, but still be useable by pps.
